From aa846d382c40809f812b2f1c7d8b92d901a75da5 Mon Sep 17 00:00:00 2001
From: Michal Sojka <michal.sojka@cvut.cz>
Date: Sat, 6 Jul 2024 22:48:54 +0200
Subject: [PATCH 1/3] Patch and fix gz-*-vendor packages for Jazzy and Rolling
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This fixes build failures in gz-*-vendor packages for Jazzy and
Rolling. Some of these packages are required for building rviz2.

Since most of the gz-* packages share the same structure, we introduce
patchGzAmentVendorGit function to make patching easier. In addition to
patching ament_vendor() calls, it adds a check to CMakeLists.txt to
detect upstream updates of the vendored package version. This should
help keeping our overrides in sync with upstream.

Currently, the overrides for Jazzy and Rolling are the same.

Co-authored-by: Błażej Sowa <bsowa123@gmail.com>
---
 distros/jazzy/overrides.nix   | 107 ++++++++++++++++++++++++++++++++++
 distros/rolling/overrides.nix | 107 ++++++++++++++++++++++++++++++++++
 lib/default.nix               |  33 +++++++++++
 3 files changed, 247 insertions(+)

diff --git a/distros/jazzy/overrides.nix b/distros/jazzy/overrides.nix
index 11a74b784e..d05a60a0af 100644
--- a/distros/jazzy/overrides.nix
+++ b/distros/jazzy/overrides.nix
@@ -31,6 +31,108 @@ in {
     fetchgitArgs.hash = "sha256-gztnxui9Fe/FTieMjdvfJjWHjkImtlsHn6fM1FruyME=";
   };
 
+  gz-cmake-vendor = lib.patchGzAmentVendorGit rosSuper.gz-cmake-vendor {
+    version = "3.5.3";
+    hash = "sha256-fnN3Fmp7F5W0JixJUEp2v/OnXzmRidS5ujmSYxIRWto=";
+  };
+
+  gz-common-vendor = (lib.patchGzAmentVendorGit rosSuper.gz-common-vendor {
+    version = "5.6.0";
+    hash = "sha256-vM+/V2F+Nr/LReqcMAmAbgAyaph/vMZVb0BO0MAUp6I=";
+  }).overrideAttrs ({
+    nativeBuildInputs ? [], ...
+  }: {
+    # https://github.com/gazebo-release/gz_common_vendor/pull/2
+    nativeBuildInputs = nativeBuildInputs ++ [ self.pkg-config ];
+  });
+
+  gz-dartsim-vendor = lib.patchAmentVendorGit rosSuper.gz-dartsim-vendor {
+    url = "https://github.com/dartsim/dart.git";
+    rev = "v6.13.2";
+    fetchgitArgs.hash = "sha256-AfKPqUiW6BsM98TIzTY2ZcFP1WvURs8/dGOzanIiB9g=";
+  };
+
+  gz-fuel-tools-vendor = lib.patchGzAmentVendorGit rosSuper.gz-fuel-tools-vendor {
+    version = "9.1.0";
+    hash = "sha256-txeIzj2vmvL5NDu6O07c7LwcCWE26OFEzvyc9TBrJAw=";
+  };
+
+  gz-gui-vendor = (lib.patchGzAmentVendorGit rosSuper.gz-gui-vendor {
+    version = "8.3.0";
+    hash = "sha256-V0zaL6qrd510hMECCr3/mMkyqf4yu2aaKLRZ6Rw0s/4=";
+  }).overrideAttrs ({
+    postInstall ? "", ...
+  }: {
+    # "RPATH of binary libGrid3D.so contains a forbidden reference to
+    # /build/" (see https://github.com/gazebosim/gz-gui/issues/627).
+    postInstall = postInstall + ''
+      ${self.patchelf}/bin/patchelf --remove-rpath $out/opt/gz_gui_vendor/lib64/gz-gui-8/plugins/libGrid3D.so
+    '';
+  });
+
+  gz-launch-vendor = lib.patchGzAmentVendorGit rosSuper.gz-launch-vendor {
+    version = "7.1.0";
+    hash = "sha256-En3V8i/Ie8+KnSHGlm9Bap7REdLhYBaVHVbOM+/Pzno=";
+  };
+
+  gz-math-vendor = lib.patchGzAmentVendorGit rosSuper.gz-math-vendor {
+    version = "7.5.0";
+    hash = "sha256-TEadejtPCR3FAUbyAAME28tmqaxypPTJDYidjZ3FPIY=";
+  };
+
+  gz-msgs-vendor = lib.patchGzAmentVendorGit rosSuper.gz-msgs-vendor {
+    version = "10.3.0";
+    hash = "sha256-PQT8EpTxafldnKG3hDSXw2P22gLRg2EfMllrzaTaDEw=";
+  };
+
+  gz-ogre-next-vendor = (lib.patchAmentVendorGit rosSuper.gz-ogre-next-vendor {
+    url = "https://github.com/OGRECave/ogre-next.git";
+    rev = "v2.3.3";
+    fetchgitArgs.hash = "sha256-elSj35LwsLzj1ssDPsk9NW/KSXfiOGYmw9hQSAWdpFM=";
+  }).overrideAttrs({ ... }: {
+    dontFixCmake = true;
+  });
+
+  gz-physics-vendor = lib.patchGzAmentVendorGit rosSuper.gz-physics-vendor {
+    version = "7.3.0";
+    hash = "sha256-PTalEQc9C/QsYMO+XK7aOzZUzC01jxiW6bjdItB5hlM=";
+  };
+
+  gz-plugin-vendor = lib.patchGzAmentVendorGit rosSuper.gz-plugin-vendor {
+    version = "2.0.3";
+    hash = "sha256-9t6vcnBbfRWu6ptmqYAhmWKDoKAaK631JD9u1C0G0mY=";
+  };
+
+  gz-rendering-vendor = lib.patchGzAmentVendorGit rosSuper.gz-rendering-vendor {
+    version = "8.2.0";
+    hash = "sha256-eaWkZKHu566Rub7YSO2lnKdj8YQbhl86v+JR4zrgtjs=";
+  };
+
+  gz-sensors-vendor = lib.patchGzAmentVendorGit rosSuper.gz-sensors-vendor {
+    version = "8.2.0";
+    hash = "sha256-j/8kS+Bvaim2gtsZcp+/u8CAE+N24/5qZhciFR0Q8+M=";
+  };
+
+  gz-sim-vendor = lib.patchGzAmentVendorGit rosSuper.gz-sim-vendor {
+    version = "8.6.0";
+    hash = "sha256-zSiPHEh3h2J8hGL342tde5U9FLaGnWs72WD9BqyPf6E=";
+  };
+
+  gz-tools-vendor = lib.patchGzAmentVendorGit rosSuper.gz-tools-vendor {
+    version = "2.0.1";
+    hash = "sha256-sV/T53oVk1fgjwqn/SRTaPTukt+vAlGGxGvTN8+G6Mo=";
+  };
+
+  gz-transport-vendor = lib.patchGzAmentVendorGit rosSuper.gz-transport-vendor {
+    version = "13.4.0";
+    hash = "sha256-2Akd3vKr07IdgoJppvUV1nZlHE4RdQfI2R18ihHTDHk=";
+  };
+
+  gz-utils-vendor = lib.patchGzAmentVendorGit rosSuper.gz-utils-vendor {
+    version = "2.2.0";
+    hash = "sha256-dNoDOZtk/zseHuOM5mOPHkXKU7wqxxKrFnh7e09bjRA=";
+  };
+
   iceoryx-hoofs = rosSuper.iceoryx-hoofs.overrideAttrs ({
     patches ? [], ...
   }: {
@@ -87,6 +189,11 @@ in {
     sha256 = "sha256-TyFt3d78GidhDGD17KgjAaZl/qvAcGJP8lmu4EOxpYg=";
   };
 
+  sdformat-vendor = lib.patchGzAmentVendorGit rosSuper.sdformat-vendor {
+    version = "14.5.0";
+    hash = "sha256-nGBLnQP0TTKDVbYGyx23Fcs79UCJveajsll2LvyLJwQ=";
+  };
+
   urdfdom = rosSuper.urdfdom.overrideAttrs ({
     patches ? [], ...
   }: {
diff --git a/distros/rolling/overrides.nix b/distros/rolling/overrides.nix
index fedb61ab11..4030fe7533 100644
--- a/distros/rolling/overrides.nix
+++ b/distros/rolling/overrides.nix
@@ -31,6 +31,108 @@ in {
     fetchgitArgs.hash = "sha256-gztnxui9Fe/FTieMjdvfJjWHjkImtlsHn6fM1FruyME=";
   };
 
+  gz-cmake-vendor = lib.patchGzAmentVendorGit rosSuper.gz-cmake-vendor {
+    version = "3.5.3";
+    hash = "sha256-fnN3Fmp7F5W0JixJUEp2v/OnXzmRidS5ujmSYxIRWto=";
+  };
+
+  gz-common-vendor = (lib.patchGzAmentVendorGit rosSuper.gz-common-vendor {
+    version = "5.6.0";
+    hash = "sha256-vM+/V2F+Nr/LReqcMAmAbgAyaph/vMZVb0BO0MAUp6I=";
+  }).overrideAttrs ({
+    nativeBuildInputs ? [], ...
+  }: {
+    # https://github.com/gazebo-release/gz_common_vendor/pull/2
+    nativeBuildInputs = nativeBuildInputs ++ [ self.pkg-config ];
+  });
+
+  gz-dartsim-vendor = lib.patchAmentVendorGit rosSuper.gz-dartsim-vendor {
+    url = "https://github.com/dartsim/dart.git";
+    rev = "v6.13.2";
+    fetchgitArgs.hash = "sha256-AfKPqUiW6BsM98TIzTY2ZcFP1WvURs8/dGOzanIiB9g=";
+  };
+
+  gz-fuel-tools-vendor = lib.patchGzAmentVendorGit rosSuper.gz-fuel-tools-vendor {
+    version = "9.1.0";
+    hash = "sha256-txeIzj2vmvL5NDu6O07c7LwcCWE26OFEzvyc9TBrJAw=";
+  };
+
+  gz-gui-vendor = (lib.patchGzAmentVendorGit rosSuper.gz-gui-vendor {
+    version = "8.3.0";
+    hash = "sha256-V0zaL6qrd510hMECCr3/mMkyqf4yu2aaKLRZ6Rw0s/4=";
+  }).overrideAttrs ({
+    postInstall ? "", ...
+  }: {
+    # "RPATH of binary libGrid3D.so contains a forbidden reference to
+    # /build/" (see https://github.com/gazebosim/gz-gui/issues/627).
+    postInstall = postInstall + ''
+      ${self.patchelf}/bin/patchelf --remove-rpath $out/opt/gz_gui_vendor/lib64/gz-gui-8/plugins/libGrid3D.so
+    '';
+  });
+
+  gz-launch-vendor = lib.patchGzAmentVendorGit rosSuper.gz-launch-vendor {
+    version = "7.1.0";
+    hash = "sha256-En3V8i/Ie8+KnSHGlm9Bap7REdLhYBaVHVbOM+/Pzno=";
+  };
+
+  gz-math-vendor = lib.patchGzAmentVendorGit rosSuper.gz-math-vendor {
+    version = "7.5.0";
+    hash = "sha256-TEadejtPCR3FAUbyAAME28tmqaxypPTJDYidjZ3FPIY=";
+  };
+
+  gz-msgs-vendor = lib.patchGzAmentVendorGit rosSuper.gz-msgs-vendor {
+    version = "10.3.0";
+    hash = "sha256-PQT8EpTxafldnKG3hDSXw2P22gLRg2EfMllrzaTaDEw=";
+  };
+
+  gz-ogre-next-vendor = (lib.patchAmentVendorGit rosSuper.gz-ogre-next-vendor {
+    url = "https://github.com/OGRECave/ogre-next.git";
+    rev = "v2.3.3";
+    fetchgitArgs.hash = "sha256-elSj35LwsLzj1ssDPsk9NW/KSXfiOGYmw9hQSAWdpFM=";
+  }).overrideAttrs({ ... }: {
+    dontFixCmake = true;
+  });
+
+  gz-physics-vendor = lib.patchGzAmentVendorGit rosSuper.gz-physics-vendor {
+    version = "7.3.0";
+    hash = "sha256-PTalEQc9C/QsYMO+XK7aOzZUzC01jxiW6bjdItB5hlM=";
+  };
+
+  gz-plugin-vendor = lib.patchGzAmentVendorGit rosSuper.gz-plugin-vendor {
+    version = "2.0.3";
+    hash = "sha256-9t6vcnBbfRWu6ptmqYAhmWKDoKAaK631JD9u1C0G0mY=";
+  };
+
+  gz-rendering-vendor = lib.patchGzAmentVendorGit rosSuper.gz-rendering-vendor {
+    version = "8.2.0";
+    hash = "sha256-eaWkZKHu566Rub7YSO2lnKdj8YQbhl86v+JR4zrgtjs=";
+  };
+
+  gz-sensors-vendor = lib.patchGzAmentVendorGit rosSuper.gz-sensors-vendor {
+    version = "8.2.0";
+    hash = "sha256-j/8kS+Bvaim2gtsZcp+/u8CAE+N24/5qZhciFR0Q8+M=";
+  };
+
+  gz-sim-vendor = lib.patchGzAmentVendorGit rosSuper.gz-sim-vendor {
+    version = "8.6.0";
+    hash = "sha256-zSiPHEh3h2J8hGL342tde5U9FLaGnWs72WD9BqyPf6E=";
+  };
+
+  gz-tools-vendor = lib.patchGzAmentVendorGit rosSuper.gz-tools-vendor {
+    version = "2.0.1";
+    hash = "sha256-sV/T53oVk1fgjwqn/SRTaPTukt+vAlGGxGvTN8+G6Mo=";
+  };
+
+  gz-transport-vendor = lib.patchGzAmentVendorGit rosSuper.gz-transport-vendor {
+    version = "13.4.0";
+    hash = "sha256-2Akd3vKr07IdgoJppvUV1nZlHE4RdQfI2R18ihHTDHk=";
+  };
+
+  gz-utils-vendor = lib.patchGzAmentVendorGit rosSuper.gz-utils-vendor {
+    version = "2.2.0";
+    hash = "sha256-dNoDOZtk/zseHuOM5mOPHkXKU7wqxxKrFnh7e09bjRA=";
+  };
+
   iceoryx-hoofs = rosSuper.iceoryx-hoofs.overrideAttrs ({
     patches ? [], ...
   }: {
@@ -82,6 +184,11 @@ in {
     '';
   };
 
+  sdformat-vendor = lib.patchGzAmentVendorGit rosSuper.sdformat-vendor {
+    version = "14.5.0";
+    hash = "sha256-nGBLnQP0TTKDVbYGyx23Fcs79UCJveajsll2LvyLJwQ=";
+  };
+
   urdfdom = rosSuper.urdfdom.overrideAttrs ({
     patches ? [], ...
   }: {
diff --git a/lib/default.nix b/lib/default.nix
index 29dc50635e..877857aa20 100644
--- a/lib/default.nix
+++ b/lib/default.nix
@@ -106,6 +106,39 @@
     '' + postPatch;
   });
 
+  # patchAmentVendorGit specialized for gz-*-vendor packages. In
+  # addition to patching ament_vendor() calls, it adds a check to
+  # CMakeLists.txt to detect upstream updates of the vendored package
+  # version.
+  patchGzAmentVendorGit = pkg: {
+    version,
+    hash,
+    tarSourceArgs ? {}
+  }: let
+    stem = lib.strings.removeSuffix "-vendor"
+      (lib.strings.removePrefix "ros-${pkg.rosDistro}-" pkg.pname); # e.g. gz-cmake
+    majorNum = lib.versions.major version;
+    patchedPkg = lib.patchAmentVendorGit pkg {
+      url = "https://github.com/gazebosim/${stem}.git";
+      originalUrl = "https://github.com/gazebosim/\${GITHUB_NAME}.git";
+      rev = "${stem}${majorNum}_${version}"; # e.g. "gz-cmake3_3.5.3"
+      fetchgitArgs.hash = hash;
+      inherit tarSourceArgs;
+    };
+  in
+    patchedPkg.overrideAttrs ({
+      pname, postPatch ? "", ...
+    }: {
+      dontFixCmake = true;      # don't replace $out/opt with $out/var/empty
+      postPatch = postPatch + ''
+        cat >> CMakeLists.txt <<'EOF'
+        if(NOT ''${LIB_VER} VERSION_EQUAL "${version}")
+          message(FATAL_ERROR "Mismatch in ${pname} version (Nix: ${version}, upstream: ''${LIB_VER}). Fix this in overrides.nix.")
+        endif()
+        EOF
+      '';
+    });
+
   patchBoostPython = pkg: pkg.overrideAttrs ({
     postPatch ? "", ...
   }: {

From fc4c0ebbe4b1d53aa2dbdca331bd74280645f961 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernd=20M=C3=BCller?= <github@muellerbernd.de>
Date: Thu, 19 Sep 2024 23:17:44 +0200
Subject: [PATCH 2/3] [add] example-ros2-gz shell

---
 examples/ros2-gz-example.nix | 42 ++++++++++++++++++++++++++++++++++++
 flake.nix                    |  1 +
 2 files changed, 43 insertions(+)
 create mode 100644 examples/ros2-gz-example.nix

diff --git a/examples/ros2-gz-example.nix b/examples/ros2-gz-example.nix
new file mode 100644
index 0000000000..82ef92d237
--- /dev/null
+++ b/examples/ros2-gz-example.nix
@@ -0,0 +1,42 @@
+# gz needs insecure packages, namely freeimage-unstable-2021-11-01
+# to run this dev shell use:
+# NIXPKGS_ALLOW_INSECURE=1 nix develop --impure .\#example-ros2-gz
+
+{pkgs ? import ../. {}}:
+with pkgs;
+with rosPackages.jazzy;
+  mkShell {
+    nativeBuildInputs = [
+      (buildEnv {
+        paths = [
+          colcon
+          ros-core
+          ament-cmake-core
+          python-cmake-module
+          # ros-gz
+
+          gz-cmake-vendor
+          gz-common-vendor
+          gz-dartsim-vendor
+          gz-fuel-tools-vendor
+          gz-gui-vendor
+          gz-launch-vendor
+          gz-math-vendor
+          gz-msgs-vendor
+          gz-ogre-next-vendor
+          gz-physics-vendor
+          gz-plugin-vendor
+          gz-rendering-vendor
+          gz-sensors-vendor
+          gz-sim-vendor
+          gz-tools-vendor
+          gz-transport-vendor
+          gz-utils-vendor
+        ];
+      })
+    ];
+    GZ_CONFIG_PATH = "${gz-tools-vendor}/opt/gz_tools_vendor/share/gz:${gz-gui-vendor}/opt/gz_gui_vendor/share/gz\:${gz-cmake-vendor}/opt/gz_cmake_vendor/share/gz\:${gz-fuel-tools-vendor}/opt/gz_fuel_tools_vendor/share/gz\:${gz-launch-vendor}/opt/gz_launch_vendor/share/gz\:${gz-math-vendor}/opt/gz_math_vendor/share/gz\:${gz-msgs-vendor}/opt/gz_msgs_vendor/share/gz\:${gz-physics-vendor}/opt/gz_physics_vendor/share/gz\:${gz-plugin-vendor}/opt/gz_plugin_vendor/share/gz\:${gz-rendering-vendor}/opt/gz_rendering_vendor/share/gz\:${gz-sensors-vendor}/opt/gz_sensors_vendor/share/gz\:${gz-sim-vendor}/opt/gz_sim_vendor/share/gz\:${gz-transport-vendor}/opt/gz_transport_vendor/share/gz\:${gz-utils-vendor}/opt/gz_utils_vendor/share/gz";
+    shellHook = ''
+      unset QT_QPA_PLATFORM
+    '';
+  }
diff --git a/flake.nix b/flake.nix
index bde6e0e277..31ac34f068 100644
--- a/flake.nix
+++ b/flake.nix
@@ -20,6 +20,7 @@
       devShells = {
         example-turtlebot3-gazebo = import ./examples/turtlebot3-gazebo.nix { inherit pkgs; };
         example-ros2-basic = import ./examples/ros2-basic.nix { inherit pkgs; };
+        example-ros2-gz = import ./examples/ros2-gz-example.nix { inherit pkgs; };
       };
     }) // {
       overlays.default = import ./overlay.nix;

From 6d04148eac0727be34e5333f6e12cfc7e86673c3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernd=20M=C3=BCller?= <github@muellerbernd.de>
Date: Fri, 20 Sep 2024 00:13:03 +0200
Subject: [PATCH 3/3] [fix] gz setup and apply suggestions from wentasah

---
 distros/jazzy/overrides.nix   | 16 ++++++++++++++--
 distros/rolling/overrides.nix | 16 ++++++++++++++--
 examples/ros2-gz-example.nix  |  1 -
 lib/default.nix               |  6 ++++++
 4 files changed, 34 insertions(+), 5 deletions(-)

diff --git a/distros/jazzy/overrides.nix b/distros/jazzy/overrides.nix
index d05a60a0af..18e88f6a3d 100644
--- a/distros/jazzy/overrides.nix
+++ b/distros/jazzy/overrides.nix
@@ -118,10 +118,22 @@ in {
     hash = "sha256-zSiPHEh3h2J8hGL342tde5U9FLaGnWs72WD9BqyPf6E=";
   };
 
-  gz-tools-vendor = lib.patchGzAmentVendorGit rosSuper.gz-tools-vendor {
+  gz-tools-vendor = (lib.patchGzAmentVendorGit rosSuper.gz-tools-vendor {
     version = "2.0.1";
     hash = "sha256-sV/T53oVk1fgjwqn/SRTaPTukt+vAlGGxGvTN8+G6Mo=";
-  };
+  }).overrideAttrs({
+    nativeBuildInputs ? [], propagatedNativeBuildInputs ? [], qtWrapperArgs ? [], postFixup ? "", ...
+  }: {
+    nativeBuildInputs = nativeBuildInputs ++ [ self.qt5.wrapQtAppsHook ];
+    propagatedNativeBuildInputs = propagatedNativeBuildInputs ++ [self.qt5.qtquickcontrols2 self.qt5.qtgraphicaleffects self.pkg-config];
+    qtWrapperArgs = qtWrapperArgs ++ [
+      # Use X11 by default
+      "--set-default QT_QPA_PLATFORM xcb"
+    ];
+    postFixup = postFixup + ''
+      wrapQtApp "$out/opt/gz_tools_vendor/bin/gz"
+    '';
+  });
 
   gz-transport-vendor = lib.patchGzAmentVendorGit rosSuper.gz-transport-vendor {
     version = "13.4.0";
diff --git a/distros/rolling/overrides.nix b/distros/rolling/overrides.nix
index 4030fe7533..143f476af0 100644
--- a/distros/rolling/overrides.nix
+++ b/distros/rolling/overrides.nix
@@ -118,10 +118,22 @@ in {
     hash = "sha256-zSiPHEh3h2J8hGL342tde5U9FLaGnWs72WD9BqyPf6E=";
   };
 
-  gz-tools-vendor = lib.patchGzAmentVendorGit rosSuper.gz-tools-vendor {
+  gz-tools-vendor = (lib.patchGzAmentVendorGit rosSuper.gz-tools-vendor {
     version = "2.0.1";
     hash = "sha256-sV/T53oVk1fgjwqn/SRTaPTukt+vAlGGxGvTN8+G6Mo=";
-  };
+  }).overrideAttrs({
+    nativeBuildInputs ? [], propagatedNativeBuildInputs ? [], qtWrapperArgs ? [], postFixup ? "", ...
+  }: {
+    nativeBuildInputs = nativeBuildInputs ++ [ self.qt5.wrapQtAppsHook ];
+    propagatedNativeBuildInputs = propagatedNativeBuildInputs ++ [self.qt5.qtquickcontrols2 self.qt5.qtgraphicaleffects self.pkg-config];
+    qtWrapperArgs = qtWrapperArgs ++ [
+      # Use X11 by default
+      "--set-default QT_QPA_PLATFORM xcb"
+    ];
+    postFixup = postFixup + ''
+      wrapQtApp "$out/opt/gz_tools_vendor/bin/gz"
+    '';
+  });
 
   gz-transport-vendor = lib.patchGzAmentVendorGit rosSuper.gz-transport-vendor {
     version = "13.4.0";
diff --git a/examples/ros2-gz-example.nix b/examples/ros2-gz-example.nix
index 82ef92d237..22288addb3 100644
--- a/examples/ros2-gz-example.nix
+++ b/examples/ros2-gz-example.nix
@@ -35,7 +35,6 @@ with rosPackages.jazzy;
         ];
       })
     ];
-    GZ_CONFIG_PATH = "${gz-tools-vendor}/opt/gz_tools_vendor/share/gz:${gz-gui-vendor}/opt/gz_gui_vendor/share/gz\:${gz-cmake-vendor}/opt/gz_cmake_vendor/share/gz\:${gz-fuel-tools-vendor}/opt/gz_fuel_tools_vendor/share/gz\:${gz-launch-vendor}/opt/gz_launch_vendor/share/gz\:${gz-math-vendor}/opt/gz_math_vendor/share/gz\:${gz-msgs-vendor}/opt/gz_msgs_vendor/share/gz\:${gz-physics-vendor}/opt/gz_physics_vendor/share/gz\:${gz-plugin-vendor}/opt/gz_plugin_vendor/share/gz\:${gz-rendering-vendor}/opt/gz_rendering_vendor/share/gz\:${gz-sensors-vendor}/opt/gz_sensors_vendor/share/gz\:${gz-sim-vendor}/opt/gz_sim_vendor/share/gz\:${gz-transport-vendor}/opt/gz_transport_vendor/share/gz\:${gz-utils-vendor}/opt/gz_utils_vendor/share/gz";
     shellHook = ''
       unset QT_QPA_PLATFORM
     '';
diff --git a/lib/default.nix b/lib/default.nix
index 877857aa20..863119136c 100644
--- a/lib/default.nix
+++ b/lib/default.nix
@@ -137,6 +137,12 @@
         endif()
         EOF
       '';
+      preBuild = ''
+        find . -name "*build.make" -print -exec sed -i "s#var/empty#opt#g" {} \;
+      '';
+      setupHook = self.writeText "${pname}-setup-hook.sh" ''
+        addToSearchPath GZ_CONFIG_PATH "@out@/opt/${lib.replaceStrings ["-"] ["_"] stem}_vendor/share/gz"
+      '';
     });
 
   patchBoostPython = pkg: pkg.overrideAttrs ({
